Implement account recovery flow and token generation

kimlisa opened the pull request:
part of rfd/0029: https://github.com/gravitational/teleport/pull/6769

#### Description
All the following checks if cloud feature is enabled, and second factor is `u2f`, `otp`, or `on`. 

Some minor changes to existing features:
- When user's get `invite` and `reset` link, 3 recovery tokens are also created, stored and returned 
- A lock for recovery attempt is added to user's resource status to separate user locked from `max failed login attempt` and `max failed recovery attempt`. This is to allow users who locked themselves from max failed login attempt to be able to still recover. And prevent a user from logging in when user max failed recovery attempt.

Recover Flow:
1.  Get recovery link
- User provides their username (only accepts emails) and one of their recovery code
- Giving 3 wrong tokens temp locks the users account (from both login and recovery attempt), and sends a locked email
- Upon valid recovery code, creates a recovery token with `subkind` set to `recovery_start` that lasts 3 hours, to indicate this is the start of recovery, and `recoverType` set to what auth cred the user wanted to recover, and emails user `/recovery/:token` link

2. Authenticate User
- User clicks on the recovery link
- User authenticates with their password or a second factor, and the request has to be accompanied by the recovery token from step 1.
- Reaching max failed recovery attempt (using same counter as step 1), deletes the recovery token and temp locks user account from login, and sends a locked email. This forces user to go back to step 1 with a new recovery code.
- On success, a new token is created with subkind set to `recovery_approved` that lasts 15 minutes, to indicate user has successfully authenticated with their password or second factor 
- Previous token is deleted

3. Change password or add a new second factor
- User provides a new password or second factor, and request must be accompanied by the new token returned from step 2
- On success,  users locks are reset so they can login right away

4. MFA related endpoints protected with a recovery approved token:
- Retrieving all mfa devices
- Deleting a device
- Getting u2f auth challenges

Rate Limiting:

Placed the new public endpoints under one rate limit set: 1 request per minute with bursts up to 10 requests, same as [auth0's change password rate limit](https://auth0.com/docs/policies/rate-limit-policy/authentication-api-endpoint-rate-limits#limits-for-production-tenants-of-paying-customers)

Account lock:
User gets locked from total of 3 invalid tries from step 1 AND 2 (not 3 tries per step)

#### Other PRs
- Enterprise endpoints: https://github.com/gravitational/teleport.e/pull/280
- Required email endpoints: https://github.com/gravitational/cloud/pull/736
- Rename ResetPasswordToken to UserToken: https://github.com/gravitational/teleport/pull/7681

kimlisa committed:
Add go-diceware lib to vendor directory

kimlisa committed:
Squash commits to merge renaming of ResetUserPassword with UserToken

russjones reviewed (commented):
@andrejtokarcik can you review this, looks like we have security sensitive changes here?

@Joerger can you review this, due to the API changes.

andrejtokarcik commented on api/types/resetpasswordtoken.go:
```diff
@@ -158,3 +175,32 @@ func (u ResetPasswordTokenV3) CheckAndSetDefaults() error {
+}
+
+// CheckSubKindType checks if given token type matches with the subkind.
+func (u *ResetPasswordTokenV3) CheckSubKindType(tokenType string) error {
```
This validation appears too ad hoc. The `tokenType` should in some way be part of `ResetPasswordToken` itself which would make these checks a matter of internal consistency that can be verified immediately upon creation of a token object. E.g., would it make sense to store the request used to create a token within the token object itself?

Or, the function could be refactored into a ```func (u *ResetPasswordTokenV3) GetType() (string, error)``` that is then checked by the caller instead of the caller passing a string in.

kimlisa replied:
I removed this check. The token spec has now a field called `RecoveryType` that stores this type of data.

andrejtokarcik reviewed (commented):
I don't like that input validation seems all over the place. It should be called within `New*` functions, unmarshalling helpers or API handlers. The loci of validation should be unified and clear. Internal logic (non-exported functions) should be able to rely on the guarantee of having valid inputs, even after the input structures happen to be modified in the future.

Also, looks like there are some linting errors, please have a look.

kimlisa commented on api/client/client.go:
```diff
@@ -576,6 +576,50 @@ func (c *Client) CreateResetPasswordToken(ctx context.Context, req *proto.Create
+	return &types.ChangePasswordWithTokenResponse{
+		WebSession:    res.GetWebSession(),
+		RecoveryCodes: res.GetRecoveryCodes(),
+	}, nil
```
revised, converted back to returning proto defined type

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,101 @@
+
+	if t.Version == "" {
+		t.Version = V1
+	}
```
revised, created a `setStaticFields` method

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,101 @@
+		Version: V1,
+		Codes:   codes,
+	}
+}
```
revised

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,101 @@
+	KindRecoverPassword             = "recover_password"
+	KindRecoverSecondFactor         = "recover_secondfactor"
+	KindRecoverPasswordApproved     = "recover_password_approved"
+	KindRecoverSecondFactorApproved = "recover_secondfactor_approved"
```
moved `KindRecoveryCodes` to `constants.go`, but the rest were deleted (this kind of information will be stored in the token now, in a field called `RecoveryType`)

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+		event.Status.UserMessage = traceErr.Error()
+
+		if err := s.emitter.EmitAuditEvent(s.closeCtx, event); err != nil {
+			log.WithFields(logrus.Fields{"user": user}).Warn("Failed to emit account recovery token used failed event.")
```
i left this unchanged b/c it produced an error, the `WithFields` param expects a type `logrus.Fields` object:

```
// Add a map of fields to the Entry.
func (entry *Entry) WithFields(fields Fields) *Entry {}

// Fields type, used to pass to `WithFields`.
type Fields map[string]interface{}
```

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+	}
+
+	if err := s.emitter.EmitAuditEvent(s.closeCtx, event); err != nil {
+		log.WithFields(logrus.Fields{"user": user}).Warn("Failed to emit account recovery token used event.")
```
unchanged, same reason as above

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+
+		if !tokenMatch && userFound {
+			traceErr = trace.BadParameter("account recovery token did not match")
+		}
```
revised to return a generic error message `invalid user or recovery code`

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+
+	log.Debugf("%v user has less than %v failed account recovery attempts", username, types.MaxRecoveryAttempts)
+
+	if err := a.UpsertRecoveryAttempt(ctx, username, attempt); err != nil {
```
n/a, I approached attempt counting similar to user logins as stated here, so this comment no longer applied: https://github.com/gravitational/teleport/pull/7405#discussion_r661718175

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+	}
+
+	attempt.Increment()
+	if attempt.Get() >= types.MaxRecoveryAttempts {
```
n/a, I approached attempt counting similar to user logins as stated here: https://github.com/gravitational/teleport/pull/7405#discussion_r661718175

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -867,6 +871,57 @@ message Events {
+    // Password is user password.
+    bytes Password = 3;
+    // U2FSignResponse is U2F sign response to a challenge.
+    U2FResponse U2FSignResponse = 4;
```
revised to `oneof`

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+	}
+
+	// Second factor must be otp, u2f, or on.
+	if authPref.GetSecondFactor() == constants.SecondFactorOff || authPref.GetSecondFactor() == constants.SecondFactorOptional {
```
revised, created a `IsSecondFactorEnforced` helper method

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,101 @@
+
+	RecoveryTokenLenBytes = 32
+	MaxRecoveryTokenTTL   = 3 * time.Hour
+	MaxRecoveryAttempts   = 3
```
moved, most were moved to `defaults.go` (tokenLen, tokenTTL, attempts), the rest went into `auth/accountrecovery.go`

kimlisa commented on lib/web/apiserver_test.go:
```diff
@@ -1474,10 +1475,13 @@ func testU2FLogin(t *testing.T, secondFactor constants.SecondFactorType) {
-		TokenID:             token.GetName(),
-		U2FRegisterResponse: u2fRegResp,
-		Password:            tempPass,
+	_, err = env.proxies[0].client.ChangePasswordWithToken(context.TODO(), &apiProto.ChangeUserAuthCredWithTokenRequest{
```
revised

kimlisa commented on lib/services/local/users.go:
```diff
@@ -1058,6 +1058,119 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
+func (s *IdentityService) DeleteRecoveryAttempt(ctx context.Context, user string) error {
+	if _, err := s.GetRecoveryAttempt(ctx, user); err != nil {
+		return trace.Wrap(err)
+	}
```
n/a (changed the attempt counting method so this no longer applied)

kimlisa commented on lib/services/local/users.go:
```diff
@@ -1058,6 +1058,119 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
+		return trace.Wrap(err)
+	}
+
+	return nil
```
n/a (changed the attempt counting method)

kimlisa commented on lib/services/local/users.go:
```diff
@@ -1058,6 +1058,119 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
+	item, err := s.Get(ctx, backend.Key(webPrefix, usersPrefix, user, recoveryAttemptsPrefix))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.NotFound("user %q is not found", user)
```
revised

kimlisa commented on lib/services/local/users.go:
```diff
@@ -1058,6 +1058,119 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
+	item := backend.Item{
+		Key:     backend.Key(webPrefix, usersPrefix, user, recoveryAttemptsPrefix),
+		Value:   value,
+		Expires: attempt.GetCreated().Add(defaults.AttemptTTL),
```
Ah i see. I decided to use the same tactic as login attempts where failure attempts are recorded per row in table. Then when i retrieve the attempts, I can place a check for each failure, to see if the attempt is past its life time.

kimlisa commented on lib/services/local/users.go:
```diff
@@ -1058,6 +1058,119 @@ func (s *IdentityService) GetGithubAuthRequest(stateToken string) (*services.Git
+	item, err := s.Get(ctx, backend.Key(webPrefix, usersPrefix, user, recoveryCodesPrefix))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.NotFound("user %q is not found", user)
```
revised

kimlisa commented on lib/services/local/resetpasswordtoken.go:
```diff
@@ -106,6 +106,31 @@ func (s *IdentityService) CreateResetPasswordToken(ctx context.Context, token ty
+func (s *IdentityService) UpdateResetPasswordToken(ctx context.Context, token types.ResetPasswordToken) error {
+	if err := token.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
```
n/a, i removed this function, b/c I realized the same counter should probably be used for the entire recovery cycle:

1) verifying recovery code: giving a wrong recovery code increases recovery attempt counter, maxing this counter in this step will temporarily lock user from further recovery attempts

2) authenticate user in token: giving wrong password or second factor increases the same counter from step 1, maxing the counter from this step will delete the token, and restart the counter, so user can restart from step 1 with a new recovery code

In both steps, users get temp locked from logging in

kimlisa commented on lib/auth/resetpasswordtoken.go:
```diff
@@ -321,11 +322,13 @@ func formatResetPasswordTokenURL(proxyHost string, tokenID string, reqType strin
+	} else if reqType == types.ResetPasswordTokenPassword {
 		u.Path = fmt.Sprintf("/web/reset/%v", tokenID)
+	} else if reqType == types.ResetPasswordTokenRecoveryStart || reqType == types.ResetPasswordTokenRecoveryApproved {
+		u.Path = fmt.Sprintf("/web/recovery/%v", tokenID)
```
revised to switch

kimlisa commented on lib/auth/password_test.go:
```diff
@@ -371,14 +372,14 @@ func (s *PasswordSuite) TestChangePasswordWithTokenErrors(c *C) {
 	c.Assert(err, IsNil)
 
-	_, err = s.a.changePasswordWithToken(context.TODO(), ChangePasswordWithTokenRequest{
+	_, err = s.a.changePasswordWithToken(context.TODO(), &proto.ChangeUserAuthCredWithTokenRequest{
```
revised

kimlisa commented on lib/auth/password_test.go:
```diff
@@ -371,14 +372,14 @@ func (s *PasswordSuite) TestChangePasswordWithTokenErrors(c *C) {
 
 	// invite token cannot be reused
-	_, err = s.a.changePasswordWithToken(context.TODO(), ChangePasswordWithTokenRequest{
+	_, err = s.a.changePasswordWithToken(context.TODO(), &proto.ChangeUserAuthCredWithTokenRequest{
```
revised

kimlisa commented on lib/auth/password_test.go:
```diff
@@ -246,7 +247,7 @@ func (s *PasswordSuite) TestChangePasswordWithToken(c *C) {
 	c.Assert(err, IsNil)
 
-	_, err = s.a.changePasswordWithToken(context.TODO(), ChangePasswordWithTokenRequest{
+	_, err = s.a.changePasswordWithToken(context.TODO(), &proto.ChangeUserAuthCredWithTokenRequest{
```
revised

kimlisa commented on lib/auth/password_test.go:
```diff
@@ -284,7 +285,7 @@ func (s *PasswordSuite) TestChangePasswordWithTokenOTP(c *C) {
 	c.Assert(err, IsNil)
 
-	_, err = s.a.changePasswordWithToken(context.TODO(), ChangePasswordWithTokenRequest{
+	_, err = s.a.changePasswordWithToken(context.TODO(), &proto.ChangeUserAuthCredWithTokenRequest{
```
revised

kimlisa commented on lib/auth/password.go:
```diff
@@ -365,6 +387,7 @@ func (s *Server) changePasswordWithToken(ctx context.Context, req ChangePassword
 
 	// Check if token exists.
 	token, err := s.GetResetPasswordToken(ctx, req.TokenID)
+	fmt.Println("------ here? ", err)
```
oops, removed

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -2741,6 +2757,88 @@ func (g *GRPCServer) GetSessionEvents(ctx context.Context, req *proto.GetSession
+func (g *GRPCServer) ChangePasswordWithToken(ctx context.Context, req *proto.ChangeUserAuthCredWithTokenRequest) (*proto.ChangePasswordWithTokenResponse, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
```
revised

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -2741,6 +2757,88 @@ func (g *GRPCServer) GetSessionEvents(ctx context.Context, req *proto.GetSession
+	sess, ok := res.WebSession.(*types.WebSessionV2)
+	if !ok {
+		err = trace.BadParameter("unexpected WebSessionV2 type %T", sess)
+		return nil, trail.ToGRPC(err)
```
n/a,i removed this, when i changed the return type back to proto defined type

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,593 @@
+		return nil, trace.Wrap(err)
+	}
+
+	if token.Expiry().Before(s.clock.Now().UTC()) {
+		return nil, trace.BadParameter("expired token")
+	}
```
I tried moving this check closer to the back where we retrieve the token from backend, but it looks this function `GetResetPasswordToken` is used in `DeleteResetPasswordToken`, and the latter relies on the getter to return an error (it broke a few tests), so I made no changes here.

kimlisa commented:
@andrejtokarcik I'm sorry to bombard you with so much changes, here is a detailed breakdown:

- There were some behavior changes so I had to rename things to better fit the behavior of the function. I am also planning to rename all `ResetPasswordToken` to a more generic `UserToken`, but I will be saving this last b/c a lot of noise was produced from renaming.

- The flow has changed that the user is now specifying what type of auth cred to recover from the beginning. If they request password, they are requesting to reset their password, if they request to recover a u2f or a totp device, then they are `adding a new mfa device` (prior it nuked all existing mfa devices). This request is now stored as `RecoverType` in the user's token.

- Prior, i used two separate attempt counter: one for verifying recovery code, another for verifying user's password or second factor. I changed this so that one counter is used in both steps.

- All other changes are from your feedback, like moving consts to proper places, adding helper funcs, placing validation checks nearer to handler

rudream commented on lib/web/ui/resetpasswordtoken.go:
```diff
@@ -28,7 +28,6 @@ type ResetPasswordToken struct {
-	// to reset secondfactor, else default is reset password.
-	IsResetSecondFactor bool `json:"isResetSecondFactor,omitempty"`
+	// RecoveryType is the recover type user requested upon starting recovery.
+	RecoveryType string `json:"recoverType,omitempty"`
```
In the request for starting the recovery process, it expects `recoveryType`:
https://github.com/gravitational/teleport.e/blob/eeb41ce45f257f10f189048868fe5a01fcab7291/lib/web/accountrecovery.go#L54-L55

But in this response it returns `recoverType`, perhaps change this to `recoveryType` to keep it consistent?

kimlisa replied:
good catch, revised

Joerger commented on api/types/types.proto:
```diff
@@ -2276,3 +2293,24 @@ message NetworkRestrictionsV4 {
         [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "spec" ];
 }
+
+// RecoveryCodes holds a user's recovery code information.
+message RecoveryCodes {
+    // Kind describes the kind of resource.
+    string Kind = 1;
+    // Version is resource version.
+    string Version = 2;
+    // Codes hold a list of recovery codes.
+    repeated RecoveryCode Codes = 3 [ (gogoproto.nullable) = false ];
+    // Created is when the recovery codes were generated.
+    google.protobuf.Timestamp Created = 4
+        [ (gogoproto.stdtime) = true, (gogoproto.nullable) = false ];
+}
```
I think `RecoveryCodes` should be converted into a proper resource since it is being stored as one. Additionally, I would update the key prefix from `web/users/[userID]/recoverycodes` to `web/recoverycodes/[userID]`, and make `RecoveryCodes.Metatdata.Name = userID`.
```suggestion
// RecoveryCodes holds a user's recovery code information.
message RecoveryCodesV1 {
    // Kind is the resource kind.
    string Kind = 1 [ (gogoproto.jsontag) = "kind" ];
    // SubKind is an optional resource subkind.
    string SubKind = 2 [ (gogoproto.jsontag) = "sub_kind,omitempty" ];
    // Version is the resource version.
    string Version = 3 [ (gogoproto.jsontag) = "version" ];
    // Metadata is the resource metadata.
    Metadata Metadata = 4 [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "metadata" ];
    // Spec is the resource spec.
    RecoveryCodesSpecV1 Spec = 5 [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "spec" ];
}

// RecoveryCodesSpecV1 is the recovery codes spec.
message RecoveryCodesSpecV1 {
    // Codes hold a list of recovery codes.
    repeated RecoveryCode Codes = 3 [ (gogoproto.nullable) = false ];
    // Created is when the recovery codes were generated.
    google.protobuf.Timestamp Created = 4
        [ (gogoproto.stdtime) = true, (gogoproto.nullable) = false ];
}
```

kimlisa replied:
converted to proper resource 

>  Additionally, I would update the key prefix from web/users/[userID]/recoverycodes to web/recoverycodes/[userID]

I kept the key format as is b/c other user resource follow this format and it is required to follow this format when we [DeleteUser](https://github.com/gravitational/teleport/blob/master/lib/services/local/users.go#L325), it deletes all user related resource by `web/users/<userID>/*`

`web/users/<userID>/params`
`web/users/<userID>/mfa/<mfaID>`
`web/users/<userID>/pwd`
`web/users/<userID>/recoverycodes`

russjones commented:
@kimlisa @alex-kovoy What's the status of this PR?

alex-kovoy commented on lib/auth/password.go:
```diff
@@ -37,19 +39,45 @@ type ChangePasswordWithTokenRequest struct {
+			return nil, trace.Wrap(err)
+		}
+		recoveryAllowed = true
+	}
```
this part is not clear, lets refactor it to something that is easier to follow:

```golang
if shouldCreateRecoveryCodes {
  //
}
```

kimlisa replied:
renamed

alex-kovoy reviewed (commented):
lets clean up the grpsServer methods by moving the validation check to the right location

alex-kovoy commented:
> @kimlisa @alex-kovoy What's the status of this PR?

@russjones  We are still working on it, just had another round of review.

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -933,6 +949,75 @@ message DeleteLockRequest {
+// token with either password or a second factor.
+message AuthenticateUserWithRecoveryTokenRequest {
+    string TokenID = 1;
+    oneof AuthCred {
```
commented

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -781,6 +781,22 @@ message GetMFADevicesRequest {}
 
+// GetMFADevicesWithTokenRequest is a request to get all the mfa devices for the user defined in
+// token.
+message GetMFADevicesWithTokenRequest { string TokenID = 1; }
```
commented

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,114 @@
+func (t *RecoveryCodes) GetKind() string               { return t.Kind }
+func (t *RecoveryCodes) GetVersion() string            { return t.Version }
+func (t *RecoveryCodes) GetCodes() []RecoveryCode      { return t.Codes }
+func (t *RecoveryCodes) SetCreation(created time.Time) { t.Created = created }
```
commented and expanded (also removed unused methods)

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,114 @@
+	Expires time.Time `json:"expires"`
+}
+
+func (a *RecoveryAttempt) Check() error {
```
This check is for the `RecoveryAttempt` struct, i was originally going to name it `CheckAndSetDefaults`, but it shouldn't be setting any default, so i left it out.

kimlisa commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// LastFailedRecoveryAttempt determines if user reached their max failed attempts.
+func LastFailedRecoveryAttempt(x int, attempts []RecoveryAttempt, now time.Time) bool {
```
renamed to `IsMaxFailedRecoveryAttempt`, and `x` to `maxAttempts`

kimlisa commented on api/types/types.proto:
```diff
@@ -848,11 +848,20 @@ message ResetPasswordTokenV3 {
 
+// RecoverType represents the types of auth credential a user can recover.
+enum RecoverType {
+    RECOVER_TOTP = 0;
```
commented

kimlisa commented on lib/auth/apiserver.go:
```diff
@@ -1153,13 +1153,26 @@ func (s *APIServer) changePasswordWithToken(auth ClientI, w http.ResponseWriter,
 	}
 
-	webSession, err := auth.ChangePasswordWithToken(r.Context(), req)
+	var u2f *proto.U2FRegisterResponse
+	if req.U2FRegisterResponse != nil {
+		u2f = &proto.U2FRegisterResponse{
+			RegistrationData: req.U2FRegisterResponse.RegistrationData,
+			ClientData:       req.U2FRegisterResponse.ClientData,
+		}
+	}
+
+	res, err := auth.ChangePasswordWithToken(r.Context(), &proto.ChangePasswordWithTokenRequest{
+		SecondFactorToken:   req.SecondFactorToken,
+		TokenID:             req.TokenID,
+		Password:            req.Password,
+		U2FRegisterResponse: u2f,
+	})
```
they are different types, so i had to convert them. `req.U2FREgisterResponse` is type `u2f.RegisterChallengeResponse` and the proto message expects type `proto.U2FRegisterResponse`

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,633 @@
+	"github.com/gravitational/trace"
+	"github.com/sethvargo/go-diceware/diceware"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/bcrypt"
```
revised

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -0,0 +1,633 @@
+}
+
+// AuthenticateUserWithRecoveryToken authenticates user defined in token with either password or second factor.
+// When a user provides a valid auth cred, the recovery token will be deleted, and an recovery approved token will be created
```
revised

kimlisa commented on lib/auth/accountrecovery_test.go:
```diff
@@ -0,0 +1,947 @@
+//  - recovery codes can be verified and marked used
+//  - reusing a used or non-existing token returns error
+func TestGenerateAndUpsertRecoveryCodes(t *testing.T) {
+	srv := newTestTLSServer(t)
```
`TestGenerateAndUpsertRecoveryCodes` seems to be the only test in the file that can be parallelized, others can't run in parallel b/c of this global variable manipulation, which the main functions check that cloud module is enabled to proceed
```
	defaultModules := modules.GetModules()
	defer modules.SetModules(defaultModules)
	modules.SetModules(&testWithCloudModules{})
```

defined here: https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L160

kimlisa commented on lib/auth/accountrecovery_test.go:
```diff
@@ -0,0 +1,947 @@
+	require.NoError(t, err)
+	require.Len(t, rc, 3)
+
+	// Test each codes are of correct format and used.
+	for _, token := range rc {
+		s := strings.Split(token, "-")
+
+		// 9 b/c 1 for prefix, 8 for words.
+		require.Len(t, s, 9)
+		require.Contains(t, token, "tele-")
+
+		// Test codes match.
+		err := srv.Auth().verifyRecoveryCode(ctx, user, []byte(token))
+		require.NoError(t, err)
+	}
+
+	// Test used codes are marked used.
+	recovery, err := srv.Auth().GetRecoveryCodes(ctx, user)
+	require.NoError(t, err)
+	for _, token := range recovery.GetCodes() {
+		require.True(t, token.IsUsed)
+	}
+
+	// Test with a used code returns error.
+	err = srv.Auth().verifyRecoveryCode(ctx, user, []byte(rc[0]))
+	require.True(t, trace.IsBadParameter(err))
+
+	// Test with invalid recoery code returns error.
+	err = srv.Auth().verifyRecoveryCode(ctx, user, []byte("invalidcode"))
+	require.True(t, trace.IsBadParameter(err))
+
+	// Test with non-existing user returns error.
+	err = srv.Auth().verifyRecoveryCode(ctx, "doesnotexist", []byte(rc[0]))
+	require.True(t, trace.IsBadParameter(err))
```
I took a look at sub-test parallelization and most of my tests are not good for it b/c most sub-tests relies on the previous actions. Like in your example, the only parallelization i can do is maybe the last two sub-tests which doesn't rely on the previous, but they aren't very expensive operations so I left it alone (but the test as a whole can be ran in parallel which i revised)

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -2868,6 +2929,123 @@ func (g *GRPCServer) DeleteLock(ctx context.Context, req *proto.DeleteLockReques
+
+	if req.GetTokenID() == "" {
+		return nil, trace.BadParameter("missing token")
+	}
```
i move these checks closer to the back

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -2868,6 +2929,123 @@ func (g *GRPCServer) DeleteLock(ctx context.Context, req *proto.DeleteLockReques
+	if req.GetPassword() == nil && req.GetSecondFactorToken() == "" && req.GetU2FRegisterResponse() == nil {
+		return nil, trace.BadParameter("missing new authentication cred")
+	}
+
```
revised

andrejtokarcik commented on lib/auth/grpcserver.go:
```diff
@@ -2899,6 +2963,90 @@ func (g *GRPCServer) DeleteLock(ctx context.Context, req *proto.DeleteLockReques
+
+	r, ok := resetToken.(*types.UserTokenV3)
+	if !ok {
+		err = trace.BadParameter("unexpected ResetPasswordToken type %T", resetToken)
+		return nil, trace.Wrap(err)
```
```suggestion
		return nil, trace.BadParameter("unexpected UserToken type %T", resetToken)
```

kimlisa replied:
revised

andrejtokarcik commented on lib/auth/password.go:
```diff
@@ -416,47 +435,112 @@ func (s *Server) changeUserSecondFactor(req ChangePasswordWithTokenRequest, toke
+}
+
+func (s *Server) createNewTOTPDevice(ctx context.Context, req newTOTPDeviceRequest) (*types.MFADevice, error) {
+	secrets, err := s.Identity.GetUserTokenSecrets(ctx, req.tokenID)
```
This query will always hit the backend. Is that desirable? Consider adding a wrapper for `GetUserTokenSecrets` (and possibly other getters you added to `IdentityService`) to `auth/auth.go`:
```go
func (a *Server) GetUserTokenSecrets(ctx context.Context, tokenID string) (types.UserTokenSecrets, error) {
	return a.GetCache().GetUserTokenSecrets(ctx, tokenID)
}
```
and then calling
```suggestion
	secrets, err := s.GetUserTokenSecrets(ctx, req.tokenID)
```
here.

kimlisa replied:
I think it depends on how quickly the cache gets updated (which i'm not sure of)

`GetUserTokenSecrets` in a happy path, is really only a one time call/use. The getter could get triggered more than once from some server error or user error (providing wrong totp token value), but this secret can also change if a user triggers `RotateUserTokenSecrets` (from like a page refresh) which upserts new secrets per call. `RotateUserTokenSecrets` gets called upon user click on a link, which then user can right away enter auth creds that trigger `GetUserTokenSecrets`.

And other getters I defined are probably not good candidates for cache either. Like `recoveryCodes` is probably going to be infrequently read, so I don't think it should be cached?  (we don't cache user password hashes and passwords will be used more often). Other getters like get `recoveryAttempts` and `getMFADevice` also relies on how quickly the cache gets updated.

andrejtokarcik commented on lib/auth/usertoken_test.go:
```diff
@@ -219,6 +220,39 @@ func TestFormatAccountName(t *testing.T) {
 	}
 }
 
+func TestCreateRecoveryToken(t *testing.T) {
+	srv := newTestTLSServer(t)
```
```suggestion
func TestCreateRecoveryToken(t *testing.T) {
	t.Parallel()
	srv := newTestTLSServer(t)
```
should work?

kimlisa replied:
revised

andrejtokarcik commented on api/types/accountrecovery.go:
```diff
@@ -0,0 +1,123 @@
+	return nil
+}
+
+// SortedRecoveryAttempts sorts recovery attempts by time.
+type SortedRecoveryAttempts []RecoveryAttempt
```
Consider making this just a private type (e.g. `recoveryAttemptsChronologically`) close to its sole use site in `services/local/users.go`. I'm not sure this is to be used so generally to warrant being exported as part of the `api` module.

kimlisa replied:
revised, moved it to file where it is being used

andrejtokarcik commented on lib/web/apiserver.go:
```diff
@@ -88,8 +88,8 @@ type Handler struct {
 
-	// clusterFeatures contain flags for supported and unsupported features.
-	clusterFeatures proto.Features
+	// ClusterFeatures contain flags for supported and unsupported features.
+	ClusterFeatures proto.Features
```
Why does this need to be exported?

kimlisa replied:
it needed to be exported to be used in the enterprise web handlers to check if cloud was enabled: https://github.com/gravitational/teleport.e/pull/280/files#diff-0c58e7b615d63a03da9676aa4c3add595d3a9e4d63f4d50990d07f6521e2d362R98

kimlisa commented on api/types/types.proto:
```diff
@@ -852,11 +852,23 @@ message UserTokenV3 {
+    RECOVER_TOTP = 0;
+    // RECOVER_U2F indicates user wants to recover a u2f device.
+    RECOVER_U2F = 1;
+    // RECOVER_PASSWORD indicates user wants to recovery their password.
```
revised

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -705,6 +705,26 @@ func (g *GRPCServer) GetResetPasswordToken(ctx context.Context, req *proto.GetRe
+
+	r, ok := token.(*types.UserTokenV3)
+	if !ok {
+		err = trace.BadParameter("unexpected UserToken type %T", token)
+		return nil, trace.Wrap(err)
```
revised

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -2899,6 +2963,90 @@ func (g *GRPCServer) DeleteLock(ctx context.Context, req *proto.DeleteLockReques
+
+	r, ok := resetToken.(*types.UserTokenV3)
+	if !ok {
+		err = trace.BadParameter("unexpected ResetPasswordToken type %T", resetToken)
+		return nil, trace.Wrap(err)
```
revised

kimlisa committed:
Address code reviews

kimlisa committed:
Fix delete marker

kimlisa committed:
Make minor changes

- Make tigther restrictions on mfa get list and mfa delete
- Delete recovery token after producing recovery codes
- Add new recovery token event to dynamic.go

kimlisa committed:
Get username for emailing

kimlisa committed:
Refactor
- Move mfa related endpoints from enterprise to oss
- Change RecoverType enum to UserTokenUsage enum for versatility
- Make recovering second factor generic, it can be
either otp or u2f

kimlisa committed:
Rename and refactor

codingllama commented:
I'm not sure I'll be able to review this in a timely manner. Are you looking for feedback in any areas in particular?

In the future, let's try to make the PRs smaller and focused in a single set of changes.

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -1171,6 +1298,11 @@ service AuthService {
     // GetMFADevices returns all MFA devices registered for the user calling
     // this RPC.
     rpc GetMFADevices(GetMFADevicesRequest) returns (GetMFADevicesResponse);
+    // CreateAuthenticationChallenge creates and returns challenges for a users MFA devices.
+    rpc CreateAuthenticationChallenge(CreateAuthenticationChallengeRequest)
+        returns (MFAAuthenticateChallenge);
+    // DeleteMFADeviceNonstream deletes a mfa device for the user defined in the token.
+    rpc DeleteMFADeviceNonstream(DeleteMFADeviceNonstreamRequest) returns (google.protobuf.Empty);
```
@codingllama

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -780,12 +786,48 @@ message DeleteMFADeviceRequestInit {
+    // Request defines how the user will be verified before receiving list of devices.
+    oneof Request {
+        // TokenID verifies existence of token and retrieves devices for user defined in token.
+        string TokenID = 1;
```
Why do we need the token here? Didn't this already work without a token?

kimlisa replied:
> Didn't this already work without a token?

Yes that was before we needed user to view their devices without being logged in. I did originally create a seperate endpoint called `GetMFADevicesWithToken` but b/c we wanted to reduce mfa api surface i combined it.

kimlisa replied:
> It escaped me that they may list devices with a token without being logged in. Is this true for all token uses? Let's get this clearly documented in all places where it is applicable.

Most of user tokens like reset, invite, and recovery are for users who are not logged in. The only exception is the privilege token (which isn't introduced until later PR)

> Also, why do we need the AuthenticatedUser field? Don't we know from the context if they are authenticated?

Revised, i saw we can get the username from context like this: https://github.com/gravitational/teleport/blob/master/lib/auth/permissions.go#L717 (without setting a default name if username is empty)

kimlisa commented on lib/auth/auth.go:
```diff
@@ -1137,6 +1143,117 @@ func (a *Server) GetMFAAuthenticateChallenge(user string, password []byte) (*MFA
+func (a *Server) DeleteMFADeviceNonstream(ctx context.Context, req *proto.DeleteMFADeviceNonstreamRequest) error {
+	token, err := a.GetUserToken(ctx, req.GetTokenID())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Restrict to certain kinds of user token and usage.
+	if !isTokenRecoveryApprovedFor2FA(token) {
+		return trace.BadParameter("invalid token")
+	}
+
+	if token.Expiry().Before(a.clock.Now().UTC()) {
+		return trace.BadParameter("expired token")
+	}
+
+	device, err := a.GetMFADevice(ctx, token.GetUser(), req.GetDeviceID())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := a.DeleteMFADevice(ctx, token.GetUser(), req.GetDeviceID()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := a.emitter.EmitAuditEvent(ctx, &apievents.MFADeviceDelete{
+		Metadata: apievents.Metadata{
+			Type: events.MFADeviceDeleteEvent,
+			Code: events.MFADeviceDeleteEventCode,
+		},
+		UserMetadata: apievents.UserMetadata{
+			User: token.GetUser(),
+		},
+		MFADeviceMetadata: mfaDeviceEventMetadata(device),
+	}); err != nil {
+		log.WithError(err).Warn("Failed to emit delete mfa device event.")
+	}
+
+	return nil
+}
+
```
@codingllama

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -780,12 +786,48 @@ message DeleteMFADeviceRequestInit {
 
+// DeleteMFADeviceNonstreamRequest is a request to delete a mfa device for the user defined in
+// token.
+message DeleteMFADeviceNonstreamRequest {
```
Maybe `DeleteMFADeviceSync`? TBH I don't like either, but can't think of anything better.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -780,12 +786,48 @@ message DeleteMFADeviceRequestInit {
+// token.
+message DeleteMFADeviceNonstreamRequest {
+    // TokenID is the ID of a user token.
+    string TokenID = 1;
```
This should be an MFAAuthenticateResponse - it should be able to call with either token or a signed challenge.

kimlisa commented on lib/auth/auth_with_roles.go:
```diff
@@ -2946,15 +2989,19 @@ func (a *ServerWithRoles) DeleteNetworkRestrictions(ctx context.Context) error {
+func (a *ServerWithRoles) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequest) (*proto.GetMFADevicesResponse, error) {
+	// Logged in users can view their own devices.
+	if req.GetTokenID() == "" {
+		req.Request = &proto.GetMFADevicesRequest_AuthenticatedUser{AuthenticatedUser: a.context.User.GetName()}
+	}
 
-// GetMFADevices exists to satisfy auth.ClientI but is not implemented here.
-// Use auth.GRPCServer.GetMFADevices or client.Client.GetMFADevices instead.
-func (a *ServerWithRoles) GetMFADevices(context.Context, *proto.GetMFADevicesRequest) (*proto.GetMFADevicesResponse, error) {
-	return nil, trace.NotImplemented("bug: GetMFADevices must not be called on auth.ServerWithRoles")
+	return a.authServer.GetMFADevices(ctx, req)
```
@codingllama

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -780,12 +786,48 @@ message DeleteMFADeviceRequestInit {
+}
+
+// UserPwd describes fields for a user's username and password.
+message UserPwd {
```
Rename to `UserCredentials`? UserPwd, to me, is just the password.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -780,12 +786,48 @@ message DeleteMFADeviceRequestInit {
+// UserPwd describes fields for a user's username and password.
+message UserPwd {
+    string Username = 1;
+    bytes Password = 2;
```
Not `string Password = 2`? Do we really expect purely binary passwords?

kimlisa replied:
It was because of this: https://github.com/gravitational/teleport/blob/master/lib/auth/password.go#L34, we convert the string password to byte for requests

kimlisa commented on lib/auth/grpcserver.go:
```diff
@@ -1944,14 +1960,34 @@ func (g *GRPCServer) GetMFADevices(ctx context.Context, req *proto.GetMFADevices
+	devs, err := actx.ServerWithRoles.GetMFADevices(ctx, req)
+	return devs, trace.Wrap(err)
+}
+
+// CreateAuthenticationChallenge creates and returns challenges for a users MFA devices.
+func (g *GRPCServer) CreateAuthenticationChallenge(ctx context.Context, req *proto.CreateAuthenticationChallengeRequest) (*proto.MFAAuthenticateChallenge, error) {
+	actx, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	res, err := actx.ServerWithRoles.CreateAuthenticationChallenge(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+// DeleteMFADeviceNonstream deletes a mfa device for the user defined in the token.
+func (g *GRPCServer) DeleteMFADeviceNonstream(ctx context.Context, req *proto.DeleteMFADeviceNonstreamRequest) (*empty.Empty, error) {
+	actx, err := g.authenticate(ctx)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	// TODO(awly): mfa: remove secrets from MFA devices.
-	return &proto.GetMFADevicesResponse{
-		Devices: devs,
-	}, nil
+
+	if err := actx.ServerWithRoles.DeleteMFADeviceNonstream(ctx, req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &empty.Empty{}, nil
 }
```
@codingllama

kimlisa reviewed (commented):
@codingllama i tagged you in the places that defines/changes mfa related endpoints

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
 
+// ChangePasswordWithTokenRequest defines a request to change one or more of a user's auth
+// credential.
+message ChangePasswordWithTokenRequest {
+    string TokenID = 1;
+    // SecondFactorToken is an otp value.
+    string SecondFactorToken = 2;
+    // Password is user's password.
+    bytes Password = 3;
+    // U2FRegisterResponse is U2F registration challenge response.
+    U2FRegisterResponse U2FRegisterResponse = 4;
+    // DeviceName is the name of the second factor device.
+    string DeviceName = 5;
+}
```
I think this should be something in the lines of the message below:

```suggestion
message ChangePasswordWithTokenRequest {
    UserCredentials Credentials = 1;
    MFAAuthenticateResponse ExistingMFAResponse = 2;
    string NewDeviceName = 3;
    MFARegisterResponse NewMFARegisterResponse = 4;
}
```

I also think this message (and RPC) are misnamed - they change both password AND add a new device, which the name doesn't imply.

kimlisa replied:
renamed rpc and mesage to `ChangeAuthentication` (did not include `MFAAuthenticateResponse` and kept TokenID for now)

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+    // Password is user's password.
+    bytes Password = 3;
+    // U2FRegisterResponse is U2F registration challenge response.
+    U2FRegisterResponse U2FRegisterResponse = 4;
```
Are the register response/new device mandatory? If not, I'd consider adding both to their own message, so it's clearer that all of this may absent.

kimlisa replied:
No it's not mandatory. I used `MFARegisterResponse` to combine secondFactorToken and u2fRegisterResponse (i think this is what you meant). I removed the `DeviceName` field b/c our UI doesn't allow users to provide their own device name on signup/reset forms (yet)

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
 
+// ChangePasswordWithTokenRequest defines a request to change one or more of a user's auth
+// credential.
+message ChangePasswordWithTokenRequest {
```
Maybe `ChangeAuthenticationRequest` and `rpc ChangeAuthentication(...)`, so it doesn't seem tied to passwords.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+
+// CreateRecoveryStartTokenRequest defines a request to create a user token for starting account
+// recovery.
+message CreateRecoveryStartTokenRequest {
```
Maybe `CreateAccountRecoveryTokenRequest`? I don't know what a "RecoveryStartToken" is, but "AccountRecovery" is easily relatable.

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+    // RecoveryCode is one of the user's recovery code.
+    bytes RecoveryCode = 2;
+    // IsRecoverPassword is a flag to indicate if request is to recover password or second factor.
+    bool IsRecoverPassword = 3;
```
Use an enum here. Avoid bools in APIs, they have terrible resilience to future changes.

https://cloud.google.com/apis/design/design_patterns#bool_vs_enum_vs_string

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+
+// SetNewAuthCredWithRecoveryTokenRequest defines a request to set either a new password or
+// add a new mfa device.
+message SetNewAuthCredWithRecoveryTokenRequest {
```
Isn't this what ChangePassword does? Let's consolidate both RPCs.

kimlisa replied:
They are slightly different. `ChangePassword` allows both a password and a second factor change (used for invite and reset tokens). This one only accepts oneof auth credential. I also renamed this to `ChangeAuthenticationFromAccountRecovery`

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+
+// AuthenticateUserWithRecoveryTokenRequest defines a request to authenticate the user defined in
+// token with either password or a second factor.
+message AuthenticateUserWithRecoveryTokenRequest {
```
This seems misnamed to me, given that AuthenticateUserWithRecoveryTokenRequest seems to return another token.

What is the use for the RPC? How is it different from CreateRecoveryStartToken?

kimlisa replied:
I renamed this to `CreateAccountRecoveryApprovedToken` with more description provided in the rpc definition

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+    // Username is the name of the user that the token belongs to.
+    string Username = 2;
+    // AuthCred is either a user's password or one of their second factor.
+    oneof AuthCred {
+        // SecondFactorToken is 2nd factor token value.
+        string SecondFactorToken = 3;
+        // Password is user password.
+        bytes Password = 4;
+        // U2FSignResponse is U2F sign response to a challenge.
+        U2FResponse U2FSignResponse = 5;
+    }
```
Similarly to others, we need to make this allow the full breadth of MFA options.

```suggestion
    oneof AuthCred {
        // Password is user password.
        string Password = 3;
        MFAAuthenticateResponse ExistingMFAResponse = 4;
    }
```

codingllama commented on api/client/proto/authservice.proto:
```diff
@@ -977,6 +1019,89 @@ message DeleteWindowsDesktopRequest {
+// defined in token.
+message CreateRecoveryCodesWithTokenRequest {
+    // TokenID is the ID of a user recovery token.
+    string TokenID = 1;
```
I'm not sure whether this should contain MFAAuthenticateResponse instead of just a token.

kimlisa replied:
MFA is not required here, just a token to protect this request b/c this request can be made without a user being logged in.

codingllama reviewed (commented):
Similarly to #7961, I am missing an overview of what are the RPCs changes we are planning here.

Ideally I'd like to see:

1. What are the new RPCs we are planning to add/replace/change and why
2. What is the lifecycle of a token, how long it lasts
3. What are the operations in which we allow tokens to be used
4. What are the operations where it is paramount that we get password+MFA before allowing access, and whether tokens may cover those

I'd also like to see an RFD for changes of this scope in the future.

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -281,6 +281,12 @@ message RotateUserTokenSecretsRequest { string TokenID = 1 [ (gogoproto.jsontag)
 message GetResetPasswordTokenRequest { string TokenID = 1 [ (gogoproto.jsontag) = "token" ]; }
 
+// GetRecoveryTokenRequest is a request to get a recovery token.
+message GetRecoveryTokenRequest {
```
To start the recovery flow, user provides their username and a valid recovery code. This creates a `recovery_start` user token, that is emailed to the user. When the user clicks on this email, browser needs to retrieve the token and the information that it stored like whether user requested to recover their password or second factor.

codingllama replied:
Ack. Let's get this info here, it's useful to know.

kimlisa commented on api/client/proto/authservice.proto:
```diff
@@ -281,6 +281,12 @@ message RotateUserTokenSecretsRequest { string TokenID = 1 [ (gogoproto.jsontag)
 // GetResetPasswordTokenRequest is a request to get a reset password token.
 message GetResetPasswordTokenRequest { string TokenID = 1 [ (gogoproto.jsontag) = "token" ]; }
 
+// GetRecoveryTokenRequest is a request to get a recovery token.
```
It is documented here: https://github.com/gravitational/teleport/pull/7405/files#diff-dc06ecd75714902724837dbeffb023dc88ec73b4eb8ca0c98ab2ec3d8c45197eR46, maybe that is not the right place to put it?

codingllama replied:
Those are way too detached from the API for someone to get to them. Let's get some of this info here as well.

kimlisa replied:
put more info on the rpc definition

codingllama commented on api/client/client.go:
```diff
@@ -1076,6 +1115,17 @@ func (c *Client) GetMFADevices(ctx context.Context, in *proto.GetMFADevicesReque
 	return resp, nil
 }
 
+// CreateAuthenticationChallenge creates and returns challenges for a users MFA devices.
+func (c *Client) CreateAuthenticationChallenge(ctx context.Context, in *proto.CreateAuthenticationChallengeRequest) (*proto.MFAAuthenticateChallenge, error) {
+	resp, err := c.grpc.CreateAuthenticationChallenge(ctx, in, c.callOpts...)
+	return resp, trail.FromGRPC(err)
+}
+
+func (c *Client) DeleteMFADeviceNonstream(ctx context.Context, in *proto.DeleteMFADeviceNonstreamRequest) error {
+	_, err := c.grpc.DeleteMFADeviceNonstream(ctx, in, c.callOpts...)
+	return trail.FromGRPC(err)
+}
+
```
I think it's marginally better, unless you have some other idea?

codingllama commented on lib/auth/auth.go:
```diff
@@ -1137,6 +1143,117 @@ func (a *Server) GetMFAAuthenticateChallenge(user string, password []byte) (*MFA
+		return nil, trace.BadParameter("missing request parameter")
+	}
+
+	protoChal, err := a.mfaAuthChallenge(ctx, username, a.Identity)
```
nit: Add a comment on the a.Identity parameter, otherwise it's not clear why it needs to be there (as mfaAuthChallenge has access to `a`).

```suggestion
	protoChal, err := a.mfaAuthChallenge(ctx, username, a.Identity /* u2fStorage */)
```

codingllama commented on lib/auth/auth.go:
```diff
@@ -1137,6 +1143,117 @@ func (a *Server) GetMFAAuthenticateChallenge(user string, password []byte) (*MFA
 	return chal, nil
 }
 
+// CreateAuthenticationChallenge creates and returns challenges for a users MFA devices.
```
Suggestion: I don't find the comments in RPC-implementing-methods important - the RPC documentation is what matters imo. Fine by me if you want to omit or do something like `// FooRPC implements AuthService.FooRPC`.

Same for others.

codingllama commented on lib/auth/auth.go:
```diff
@@ -1137,6 +1143,117 @@ func (a *Server) GetMFAAuthenticateChallenge(user string, password []byte) (*MFA
+		}
+
+		// Restrict to certain kinds of user token and usage.
+		if !isTokenRecoveryApprovedFor2FA(token) {
+			return nil, trace.BadParameter("invalid token")
+		}
+
+		if token.Expiry().Before(a.clock.Now().UTC()) {
+			return nil, trace.BadParameter("expired token")
+		}
```
Suggestion: push this logic into the token class, or refactor to an utility method?

Eg:

```go
err := token.Verify(allowedUsages, currentTime)
// verify err
```

codingllama commented on lib/auth/auth_with_roles.go:
```diff
@@ -2946,15 +2989,19 @@ func (a *ServerWithRoles) DeleteNetworkRestrictions(ctx context.Context) error {
+func (a *ServerWithRoles) GetMFADevices(ctx context.Context, req *proto.GetMFADevicesRequest) (*proto.GetMFADevicesResponse, error) {
+	// Logged in users can view their own devices.
+	if req.GetTokenID() == "" {
+		req.Request = &proto.GetMFADevicesRequest_AuthenticatedUser{AuthenticatedUser: a.context.User.GetName()}
```
Let's take the `AuthenticatedUser` out of the public API and pass it downwards in the stack through other means (using the context, for example).

codingllama commented on lib/auth/grpcserver.go:
```diff
@@ -1944,14 +1960,34 @@ func (g *GRPCServer) GetMFADevices(ctx context.Context, req *proto.GetMFADevices
 	}
 
-	devs, err := actx.authServer.GetMFADevices(ctx, actx.User.GetName())
+	// TODO(awly): mfa: remove secrets from MFA devices.
```
Do you know Andrew meant here? Is it the key from TOTP devices?

If yes, then we should clear those out, specially since now we're adding new ways to list devices.

kimlisa replied:
> Do you know Andrew meant here? Is it the key from TOTP devices?

Double checked with andrew and he said yes and that u2f device data we store is not secret, regardless of this, the field that stores secret for TOTP is `Device` for both totp and u2f, so I set them both to nil if request is made without secrets. I used the same `withSecrets bool` approach as others like `GetUsers` and `GetSamlConnector`.

kimlisa committed:
Address code reviews part 1

kimlisa committed:
Address CR Part 2: Add option to retrive mfa devices without secrets

kimlisa commented:
closing because this PR is being broken into smaller PRs
